src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,239 @@
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(context =>
+            {
+                var eventArgsTypeSymbol = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemEventArgs);
+
+                context.RegisterOperationBlockStartAction(context => AnalyzeOperationBlockStart(context, eventArgsTypeSymbol));
+            });
+        }
+
+        private static void AnalyzeOperationBlockStart(OperationBlockStartAnalysisContext context, INamedTypeSymbol? eventArgsTypeSymbol)
+        {
+            if (context.OwningSymbol is not IMethodSymbol methodSymbol ||
+                methodSymbol.IsOverride ||
+                methodSymbol.Parameters.Length == 0 ||
+                methodSymbol.IsImplementationOfAnyInterfaceMember() ||
+                methodSymbol.HasEventHandlerSignature(eventArgsTypeSymbol) ||
+                !context.Options.MatchesConfiguredVisibility(Rule, methodSymbol, context.Compilation,
+                    defaultRequiredVisibility: SymbolVisibilityGroup.All) ||
+                context.Options.IsConfiguredToSkipAnalysis(Rule, methodSymbol,
+                    context.Compilation))
+            {
+                return;
+            }
+
+            var typesUsedPerParameter = PooledConcurrentDictionary<IParameterSymbol, PooledConcurrentSet<ITypeSymbol>>.GetInstance(SymbolEqualityComparer.Default);
+            foreach (var param in methodSymbol.Parameters)
+            {
+                typesUsedPerParameter.AddOrUpdate(param, PooledConcurrentSet<ITypeSymbol>.GetInstance(SymbolEqualityComparer.Default), (param, value) => value);
+            }
+
+            context.RegisterOperationAction(context =>
+            {
+                var paramReference = (IParameterReferenceOperation)context.Operation;
+
+                switch (paramReference.Parent.Kind)
+                {
+                    case OperationKind.Invocation:
+                        AddDeclaringTypeSymbol(((IInvocationOperation)paramReference.Parent).TargetMethod);
+                        break;
+
+                    case OperationKind.PropertyReference:
+                        AddDeclaringTypeSymbol(((IPropertyReferenceOperation)paramReference.Parent).Property);
+                        break;
+
+                    case OperationKind.EventReference:
+                        AddDeclaringTypeSymbol(((IEventReferenceOperation)paramReference.Parent).Event);
+                        break;
+
+                    case OperationKind.FieldReference:
+                        AddType(((IFieldReferenceOperation)paramReference.Parent).Field.ContainingType);
+                        break;
+
+                    case OperationKind.MethodReference:
+                        AddType(((IMethodReferenceOperation)paramReference.Parent).Method.ContainingType);
+                        break;
```

Comment:
All those `IOperation`s are `IMemberReferenceOperation`, could we group them under a single `case` and use `AddDeclaringTypeSymbol(((IMemberReferenceOperation)paramReference.Parent).Member)`. I'm not sure if `AddDeclaringTypeSymbol` would do the right thing with this suggestion. But something to consider if it can be done.